File: magmac/src/java/magma/api/contain/stream/FlatMapHead.java
package magma.api.contain.stream;

import magma.api.option.None;
import magma.api.option.Option;

import java.util.function.Function;

class FlatMapHead<T, R> implements Head<R> {
    private final Head<T> outer;
    private final Function<T, Head<R>> mapper;
    private Head<R> current;

    public FlatMapHead(Head<R> initial, HeadedStream<T> outer, Function<T, Head<R>> mapper) {
        this.outer = outer;
        this.mapper = mapper;
        current = initial;
    }

    @Override
    public Option<R> head() {
        while (true) {
            var currentHead = current.head();
            if (currentHead.isPresent()) return currentHead;

            var tuple = outer.head()
                    .map(mapper)
                    .toTuple(current);

            if (tuple.left()) {
                current = tuple.right();
            } else {
                return None.None();
            }
        }
    }
}

File: src/windows/magma/api/collect/head/FlatMapHead.c
#include "./FlatMapHead.h"
export class FlatMapHead<T, R> implements Head<R> {
	(arg0 : T) => Query<R> mapper;
	Head<T> head;
	Query<R> current;
}

constructor (Head<T> head, Query<R> initial, (arg0 : T) => Query<R> mapper) {
	this.head = head;
	this.current = initial;
	this.mapper = mapper;
}
Option<R> next() {
	while (true) {
		Option<R> next = this.current.next();
		if (next.isPresent()) {
			return next;
		}
		Tuple2<Bool, Query<R>> tuple = this.head.next().map(this.mapper).toTuple(this.current);
		if (tuple.left()) {
			this.current = tuple.right();
		}
		else {
			return new None<R>();
		}
	}
}